### PR TITLE
[ONNX] Add quantization support to more single output ops

### LINKED ISSUE #80039: [ONNX] Increase coverage of quantization operators
Up-to-date supported list: https://pytorch.org/docs/master/onnx_supported_aten_ops.html?highlight=onnx

------

Extending support for operators used by quantization. A few thoughts/options.

1. Support all operators documented here https://pytorch.org/docs/stable/quantization-support.html?highlight=quantization. The list is non-exhaustive and there may exist undocumented quantized operators.
    * torch
        - [x] quantize_per_tensor
        - [x] quantize_per_channel
        - [x] dequantize

    * other aten
        - [ ] slice

    * torch.Tensor
        - [x] view
        - [x] as_strided
        - [x] expand
        - [x] flatten
        - [x] select
        - [x] ne
        - [x] eq
        - [x] ge
        - [x] le
        - [x] gt
        - [x] lt
        - [ ] copy_
        - [x] clone
        - [ ] dequantize
        - [ ] equal
        - [ ] int_repr
        - [ ] max
        - [x] mean
        - [ ] min
        - [ ] q_scale
        - [ ] q_zero_point
        - [ ] q_per_channel_scales
        - [ ] q_per_channel_zero_points
        - [ ] q_per_channel_axis
        - [ ] ~resize_~ [resize_ is not traceable]
        - [ ] sort
        - [ ] topk

    * torch.nn.quantized
        - [x] ReLU6
        - [x] Hardswish
        - [ ] ELU
        - [x] LeakyReLU
        - [x] Sigmoid
        - [ ] BatchNorm2d
        - [ ] BatchNorm3d
        - [ ] Conv1d
        - [x] Conv2d
        - [ ] Conv3d
        - [ ] ConvTranspose1d
        - [ ] ConvTranspose2d
        - [ ] ConvTranspose3d
        - [x] Embedding
        - [x] EmbeddingBag
        - [ ] FloatFunctional
        - [ ] FXFloatFunctional
        - [ ] QFunctional
        - [x] Linear
        - [x] LayerNorm
        - [x] GroupNorm
        - [x] InstanceNorm1d
        - [x] InstanceNorm2d
        - [x] InstanceNorm3d

    * torch.nn.quantized.functional
        - [ ] avg_pool2d
        - [ ] avg_pool3d
        - [ ] adaptive_avg_pool2d
        - [ ] adaptive_avg_pool3d
        - [ ] conv1d
        - [x] conv2d
        - [ ] conv3d
        - [ ] interpolate
        - [x] linear
        - [ ] max_pool1d
        - [ ] max_pool2d
        - [ ] celu
        - [x] leaky_relu
        - [x] hardtanh
        - [x] hardswish
        - [ ] threshold
        - [x] elu
        - [x] hardsigmoid
        - [ ] clamp
        - [ ] upsample
        - [ ] upsample_bilinear
        - [ ] upsample_nearest
1. Support all existing supported full precision pytorch/vision models with quantization.
1. Automate the discovery (support) for quantized operators.

### LINKED ISSUE #81566: [ONNX] Improve quantization error message
1. When the exporter sees unsupported quantized ops, the current message is around "unsupported TupleConstruct," because we use a tuple construct to wrap a quantized tensor. We should improve the message such that users knows they are seeing an unsupported quantized op.
2. When a symbolic function is not wrapped with the `quantized_args` wrapper, we get
    ```
      File "/dev/pytorch/torch/onnx/utils.py", line 1419, in _export
        graph, params_dict, torch_out = _model_to_graph(
      File "/dev/pytorch/torch/onnx/utils.py", line 1084, in _model_to_graph
        _C._jit_pass_onnx_assign_output_shape(
    RuntimeError: ArrayRef: invalid index Index = 1; Length = 1
    ```

### LINKED ISSUE #81678: Export quantized shufflenet_v2_x0_5 to ONNX
### 🐛 Describe the bug

Fail to export quantized shufflenet_v2_x0_5 to ONNX using the following code:

```python
import io
import numpy as np

import torch
import torch.utils.model_zoo as model_zoo
import torch.onnx

import torchvision.models.quantization as models

torch_model = models.shufflenet_v2_x0_5(pretrained=True, quantize=True)
torch_model.eval()

batch_size = 1
input_float = torch.zeros(1, 3, 224, 224)
input_tensor = input_float

torch_out = torch.jit.trace(torch_model, input_tensor)

# Export the model
torch.onnx.export(torch_model,               # model being run
                  input_tensor,              # model input (or a tuple for multiple inputs)
                  "shufflenet_v2_x0_5_qdq.onnx",   # where to save the model (can be a file or file-like object)
                  export_params=True,        # store the trained parameter weights inside the model file
                  opset_version=16,          # the ONNX version to export the model to
                  do_constant_folding=True,  # whether to execute constant folding for optimization
                  input_names = ['input'],   # the model's input names
                  output_names = ['output'], # the model's output names
                  dynamic_axes={'input' : {0 : 'batch_size'},    # variable length axes
                                'output' : {0 : 'batch_size'}})
```

Stacktrace:

```text
/home/skyline/miniconda3/envs/torch-nightly-py39/lib/python3.9/site-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and will be removed in 0.15, please use 'weights' instead.
  warnings.warn(
/home/skyline/miniconda3/envs/torch-nightly-py39/lib/python3.9/site-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and will be removed in 0.15. The current behavior is equivalent to passing `weights=ShuffleNet_V2_X0_5_QuantizedWeights.IMAGENET1K_FBGEMM_V1`. You can also use `weights=ShuffleNet_V2_X0_5_QuantizedWeights.DEFAULT` to get the most up-to-date weights.
  warnings.warn(msg)
/home/skyline/miniconda3/envs/torch-nightly-py39/lib/python3.9/site-packages/torch/onnx/_patch_torch.py:67: UserWarning: The shape inference of prim::TupleConstruct type is missing, so it may result in wrong shape inference for the exported graph. Please consider adding it in symbolic function. (Triggered internally at /opt/conda/conda-bld/pytorch_1658041842671/work/torch/csrc/jit/passes/onnx/shape_type_inference.cpp:1874.)
  torch._C._jit_pass_onnx_node_shape_type_inference(
Traceback (most recent call last):
  File "/home/skyline/Projects/python-playground/ml/ModelConversion/torch_quantizied_model_to_onnx.py", line 20, in <module>
    torch.onnx.export(torch_model,               # model being run
  File "/home/skyline/miniconda3/envs/torch-nightly-py39/lib/python3.9/site-packages/torch/onnx/utils.py", line 479, in export
    _export(
  File "/home/skyline/miniconda3/envs/torch-nightly-py39/lib/python3.9/site-packages/torch/onnx/utils.py", line 1411, in _export
    graph, params_dict, torch_out = _model_to_graph(
  File "/home/skyline/miniconda3/envs/torch-nightly-py39/lib/python3.9/site-packages/torch/onnx/utils.py", line 1054, in _model_to_graph
    graph = _optimize_graph(
  File "/home/skyline/miniconda3/envs/torch-nightly-py39/lib/python3.9/site-packages/torch/onnx/utils.py", line 624, in _optimize_graph
    graph = _C._jit_pass_onnx(graph, operator_export_type)
  File "/home/skyline/miniconda3/envs/torch-nightly-py39/lib/python3.9/site-packages/torch/onnx/utils.py", line 1744, in _run_symbolic_function
    return symbolic_fn(g, *inputs, **attrs)
  File "/home/skyline/miniconda3/envs/torch-nightly-py39/lib/python3.9/site-packages/torch/onnx/symbolic_opset13.py", line 716, in conv2d_relu
    input, input_scale, _, _ = symbolic_helper.dequantize_helper(g, q_input)
  File "/home/skyline/miniconda3/envs/torch-nightly-py39/lib/python3.9/site-packages/torch/onnx/symbolic_helper.py", line 1235, in dequantize_helper
    unpacked_qtensors = _unpack_tuple(qtensor)
  File "/home/skyline/miniconda3/envs/torch-nightly-py39/lib/python3.9/site-packages/torch/onnx/symbolic_helper.py", line 167, in _unpack_tuple
    raise RuntimeError(
RuntimeError: ONNX symbolic expected node type `prim::TupleConstruct`, got `%423 : Tensor(*, *, *, *) = onnx::Slice(%x.8, %419, %422, %411), scope: __module.stage2/__module.stage2.1 # /home/skyline/miniconda3/envs/torch-nightly-py39/lib/python3.9/site-packages/torchvision/models/quantization/shufflenetv2.py:42:0
`
```

### Versions

PyTorch version: 1.13.0.dev20220717
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04 LTS (x86_64)
GCC version: (Ubuntu 11.2.0-19ubuntu1) 11.2.0
Clang version: 14.0.6
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.9.12 (main, Jun  1 2022, 11:38:51)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.10.102.1-microsoft-standard-WSL2-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.1
[pip3] torch==1.13.0.dev20220717
[pip3] torchaudio==0.13.0.dev20220717
[pip3] torchvision==0.14.0.dev20220717
[conda] blas                      1.0                         mkl
[conda] cpuonly                   2.0                           0    pytorch-nightly
[conda] cudatoolkit               10.2.89              hfd86e86_1
[conda] mkl                       2021.4.0           h06a4308_640
[conda] mkl-service               2.4.0            py39h7f8727e_0
[conda] mkl_fft                   1.3.1            py39hd3c417c_0
[conda] mkl_random                1.2.2            py39h51133e4_0
[conda] numpy                     1.23.1                   pypi_0    pypi
[conda] numpy-base                1.22.3           py39hf524024_0
[conda] pytorch                   1.13.0.dev20220717     py3.9_cpu_0    pytorch-nightly
[conda] pytorch-mutex             1.0                         cpu    pytorch-nightly
[conda] torchaudio                0.13.0.dev20220717        py39_cpu    pytorch-nightly
[conda] torchvision               0.14.0.dev20220717        py39_cpu    pytorch-nightly

### PR DESCRIPTION
**Pull request by justinchuby**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #83008

#80039

- Implement quantization support for single output ops 
  - quantized::sigmoid
  - quantized::instance_norm
  - aten::reshape
  - aten::reshape_as
  - aten::sum
  - aten::mean
  - aten::prod
  - aten::t
  - aten::numpy_T
  - aten::expand
  - aten::expand_as
  - aten::embedding
  - aten::embedding_bag
  - aten::view
  - aten::select
  - aten::eq
  - aten::ne
  - aten::gt
  - aten::lt
  - aten::le
  - aten::ge
  - quantized::layer_norm
  - aten::elu
  - aten::selu
  - aten::maximum
  - aten::minimum
  - aten::amax
  - aten::amin
  - aten::hardtanh
  - aten::hardswish
  - quantized::group_norm
  - aten::as_strided
  - quantized::leaky_relu
  - aten::transpose
- Avoid modifying functions in `quantized_args` and have the wrapper closed over `scale` and `zero_point` instead (for purity)
- Remove magic number and assign it to INT64_MAX
- implement `_unpack_quantized_tensor` for handling quantized tensor unpacking to separate the logic from tuple unpacking and for clearer error handling

### CODE REVIEW

**Comment by justinchuby** [comment]
Commit message:

- Remove magic number and assign it to INT64_MAX
- implement `_unpack_quantized_tensor` for handling quantized tensor unpacking to separate the logic from tuple unpacking
-

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 1]
```diff
@@ -11748,26 +11748,104 @@ def test_quantized_conv2d_relu(self):
...
-    @skipIfUnsupportedMinOpsetVersion(10)
-    def test_quantized_hardsigmoid(self):
-        model = torch.nn.Hardsigmoid()
-        input = torch.randn(2, 6)
-        q_input = torch.quantize_per_tensor(input, 0.26, 128, torch.quint8)
-        self.run_test(model, q_input)
+        class Model(torch.nn.Module):
+            def __init__(self, function):
+                super().__init__()
+                self.function = function
```
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:11843]: `function` is not accurate in most cases, where it is `module` or `sub_module`.

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 2]
```diff
@@ -11748,26 +11748,104 @@ def test_quantized_conv2d_relu(self):
...
         self.run_test(model, q_input)
 
+    @common_utils.parametrize(
+        "function",
+        [
+            common_utils.subtest(
+                torch.nn.ReLU(),
+                name="relu",
+            ),
+            # common_utils.subtest(
```
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:11758]: Is this still in progress? Perhaps better to either skip with reason or remove from this PR.

File: torch/onnx/symbolic_helper.py [hunk 3]
```diff
@@ -348,44 +369,55 @@ def q_foo(g, x, y):
...
                     if _zero_point is None:
-                        _zero_point = zero_point
+                        _zero_point = arg_zero_point
                 else:
-                    dequantized_args.append(arg)
+                    maybe_dequantized_args.append(arg)
             # TODO(justinchuby): Only single output is supported for now. We may want to
             # support multiple outputs in the future.
-            output = fn(g, *dequantized_args, **kwargs)
+            output = fn(g, *maybe_dequantized_args, **kwargs)
```
Comment (BowenBao) [inline torch/onnx/symbolic_helper.py:415]: What is the rationale behind the renaming. Is empty list part of expected/accepted outcome?

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 2 above)
Comment (justinchuby) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:11758]: It is a WIP for preliminary reviews. There are four tests I needed to fix

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 1 above)
Comment (justinchuby) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:11865]: Updated to function_or_module. What do you think?

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 2 above)
Comment (justinchuby) [reply test/onnx/test_pytorch_onnx_onnxruntime.py]: Done

File: torch/onnx/symbolic_helper.py (hunk 3 above)
Comment (justinchuby) [reply torch/onnx/symbolic_helper.py:416]: The thinking was that some ops were originally non-quantized and thus not "dequantized." Thoughts?

**Comment by justinchuby** [comment]
@BowenBao ptal

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 1 above)
Comment (BowenBao) [reply test/onnx/test_pytorch_onnx_onnxruntime.py]: sgtm

File: torch/onnx/symbolic_helper.py (hunk 3 above)
Comment (BowenBao) [reply torch/onnx/symbolic_helper.py:416]: Perhaps `unquantized_args` or `nonquantized_args` is better? The `maybe` word gave me a feeling that there are some quantized args that were unable to get dequantized. IIUC, the args should all become not quantized at this point, hence the mentioned names (or the original name) in the beginning of this comment.

File: torch/onnx/symbolic_opset10.py [hunk 4]
```diff
@@ -610,6 +610,82 @@ def hardswish(g, x, op_scale, op_zero_point):
...
+        q_input,
+        weight,
+        bias,
+        eps,
+        op_scale,
+        op_zero_point,
+    ):
+        input, input_scale, _, _ = symbolic_helper.dequantize_helper(g, q_input)
+
+        # Duplicated from opset9.instance_norm
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset10.py:658]: Why not call the function directly? If that doesn't work please add the reasoning into comment too.

**Comment by BowenBao** [comment]
Please checkout CI failure

File: torch/onnx/symbolic_helper.py (hunk 3 above)
Comment (justinchuby) [reply torch/onnx/symbolic_helper.py]: changed to non_quantized_args

File: torch/onnx/symbolic_opset10.py (hunk 4 above)
Comment (justinchuby) [reply torch/onnx/symbolic_opset10.py]: There wasn't a good reason. Updated to call the function.

**Review by BowenBao — APPROVED** [review]
🚀

**Pull request closed without merging**